src/upgrades/v0_9_0/pubd_objects_migration.rs

```diff
@@ -46,28 +46,42 @@ impl PubdObjectsMigration {
         Handle::from_str(PUBSERVER_DFLT).unwrap()
     }
 
-    pub fn migrate(config: Arc<Config>) -> UpgradeResult<()> {
-        let store = KeyValueStore::disk(&config.data_dir, PUBSERVER_DIR)?;
-        let new_store = AggregateStore::disk(&config.data_dir, PUBSERVER_DIR)?;
+    pub fn prepare(mode: UpgradeMode, config: Arc<Config>) -> UpgradeResult<()> {
+        let upgrade_data_dir = config.upgrade_data_dir();
+        let current_kv_store = KeyValueStore::disk(&config.data_dir, PUBSERVER_DIR)?;
+        let new_kv_store = KeyValueStore::disk(&upgrade_data_dir, PUBSERVER_DIR)?;
+        let new_agg_store = AggregateStore::disk(&upgrade_data_dir, PUBSERVER_DIR)?;
 
-        let store_migration = PubdStoreMigration { store, new_store };
+        let store_migration = PubdStoreMigration {
+            current_kv_store,
+            new_kv_store,
+            new_agg_store,
+        };
 
         if store_migration.needs_migrate()? {
-            info!("Krill will now migrate your existing Publication Server data to the 0.9 format");
+            info!("Migrate the existing Publication Server data to the 0.9 format");
             Self::populate_repo_content(config)?;
-            store_migration.migrate()
+            store_migration.prepare_new_data(mode)
         } else {
             Ok(())
         }
     }
 
+    /// Populate the 0.9.x style repository content. Overwrite any existing content if it
+    /// exists - this is expected if the operators used "prepare-upgrade" and we are now
+    /// resuming the migration.
+    ///
+    /// NOTE: this code is not called if the migration to 0.9.x would be complete.
     fn populate_repo_content(config: Arc<Config>) -> UpgradeResult<()> {
+        info!("Populate the repository content based on current state");
         let old_store = AggregateStore::<OldRepository>::disk(&config.data_dir, PUBSERVER_DIR)?;
         if old_store.warm().is_err() {
-            // this is most likely because the info last event is off by one, try deleting the info
+            // This is most likely because the info last event is off by one which affected
+            // some old version. Try archiving the info.json to force that the publication
```

Comment:
Ah maybe this references the off-by-one comment I made above?

Reply:
no - this an off by one in versions I have forgotten about already. The info file is a helper but krill can work it out without it.